File: backtrack/GenerateParenthesis.cs
public class Solution {
     public void backtrack(IList<string> result, string str, int open, int close, int max) {
        if(str.Length == max*2){
            result.Add(str);
            return;
        }     
        if(open < max)
            backtrack(result, str+"(", open+1, close, max);
        if(close < open)
            backtrack(result, str+")", open, close+1, max);
    }
    public IList<string> GenerateParenthesis(int n) {
        IList<string> result = new List<string>();
        backtrack(result, "", 0, 0, n);
        return result;
    }
}

File: backtrack/generateParenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def backtrack(leftRemain, rightRemain, path, result):
            if leftRemain > rightRemain or leftRemain < 0 or rightRemain < 0:
                return  # backtracking
            if leftRemain == 0 and rightRemain == 0:
                result.append(path)
                return 
            backtrack(leftRemain-1, rightRemain, path+"(", result)
            backtrack(leftRemain, rightRemain-1, path+")", result)
        result = []
        backtrack (n, n, "", result)
        return result